server/src/main/java/org/elasticsearch/common/xcontent/ChunkedToXContentBuilder.java

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.xcontent;
+
+import org.elasticsearch.common.CheckedBiConsumer;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ToXContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+
+/**
+ * A fluent builder to create {@code Iterator&lt;ToXContent&gt;} objects
+ */
+public class ChunkedToXContentBuilder implements Iterator<ToXContent> {
+
+    public static ChunkedToXContentBuilder builder(ToXContent.Params params) {
+        return new ChunkedToXContentBuilder(params);
+    }
+
+    private final ToXContent.Params params;
+    private final Stream.Builder<ToXContent> builder = Stream.builder();
+    private Iterator<ToXContent> iterator;
+
+    public ChunkedToXContentBuilder(ToXContent.Params params) {
+        this.params = params;
+    }
+
+    public ChunkedToXContentBuilder startObject() {
+        builder.add((b, p) -> b.startObject());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder startObject(String name) {
+        builder.add((b, p) -> b.startObject(name));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder endObject() {
+        builder.add((b, p) -> b.endObject());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder object(String name, Consumer<ChunkedToXContentBuilder> contents) {
+        return startObject(name).execute(contents).endObject();
+    }
+
+    public ChunkedToXContentBuilder startArray() {
+        builder.add((b, p) -> b.startArray());
+        return this;
+    }
+
+    public ChunkedToXContentBuilder startArray(String name) {
+        builder.add((b, p) -> b.startArray(name));
+        return this;
+    }
+
+    public ChunkedToXContentBuilder endArray() {
+        builder.add((b, p) -> b.endArray());
+        return this;
+    }
```

Comment:
Likewise here, can we introduce an `.array()` utility that makes sure the output is well-formed?

Reply:
There already is a basic one - so these can go completely

Reply:
They're still there, and the basic one only works for strings. Almost every array is going to be made from `.startArray(n).forEach(i, f).endArray()`, let's make this one method.

Reply:
Done - they're now private methods